src/engine/PDKind.cc

```diff
@@ -0,0 +1,335 @@
+#include <set>
+#include <tuple>
+
+VerificationResult PDKind::solve(ChcDirectedHyperGraph const & graph) {
+    auto pipeline = Transformations::towardsTransitionSystems();
+    auto transformationResult = pipeline.transform(std::make_unique<ChcDirectedHyperGraph>(graph));
+    auto transformedGraph = std::move(transformationResult.first);
+    auto translator = std::move(transformationResult.second);
+    if (transformedGraph->isNormalGraph()) {
+        auto normalGraph = transformedGraph->toNormalGraph();
+        auto res = solve(*normalGraph);
+        return computeWitness ? translator->translate(std::move(res)) : std::move(res);
+    }
+    return VerificationResult(VerificationAnswer::UNKNOWN);
+}
+
+VerificationResult PDKind::solve(ChcDirectedGraph const & system) {
+    if (isTrivial(system)) {
+        return solveTrivial(system);
+    }
+    if (isTransitionSystem(system)) {
+        auto ts = toTransitionSystem(system);
+        return solveTransitionSystem(*ts);
+    }
+    SingleLoopTransformation transformation;
+    auto[ts, backtranslator] = transformation.transform(system);
+    assert(ts);
+    auto res = solveTransitionSystem(*ts);
+    return res;
+}
+
+/**
+ * Solve system with PDKIND algorithm.
+ * @param system
+ * @return
+ */
+VerificationResult PDKind::solveTransitionSystem (TransitionSystem const & system) {
+    PTRef init = system.getInit();
+    PTRef query = system.getQuery();
+
+    ReachabilityChecker reachability_checker(logic, system);
+
+    { // Check for system with empty initial states
+        SMTConfig config;
+        TimeMachine tm{logic};
+        MainSolver init_solver(logic, config, "Empty initial states");
+        init_solver.insertFormula(init);
+        auto res = init_solver.check();
+        if (res == s_False) { return VerificationResult{VerificationAnswer::SAFE}; }
+
+        init_solver.insertFormula(query);
+        res = init_solver.check();
+        if (res == s_True) { return VerificationResult{VerificationAnswer::UNSAFE}; }
+    }
+    
+    int n = 0;
+    PTRef p = logic.mkNot(query);
+    InductionFrame inductionFrame;
+    inductionFrame.insert(IFrameElement(p, logic.mkNot(p)));
+
+    // Solve the system by iteratively trying to construct an inductive strengthening of (p, not p) induction frame.
+    while (true) {
+        int k = n + 1; /* Pick k such that 1 <= k <= n + 1 */
+        auto res = push(system, inductionFrame, n, k, reachability_checker);
+
+        if (res.is_invalid) {
+            return VerificationResult(VerificationAnswer::UNSAFE, InvalidityWitness{});
+        }
+        if (res.i_frame == res.new_i_frame) {
+            return VerificationResult(VerificationAnswer::SAFE);
+        }
+        n = res.n;
+        inductionFrame = res.new_i_frame;
+    }
+}
+
+/**
+ * Check if p is invariant by iteratively constructing k-inductive strengthening of p.
+ * @param system
+ * @param iframe
+ * @param n
+ * @param k
+ * @param reachability_checker
+ * @return
```

Comment:
Either explain the parameters and the return value or remove `@param` and `@return`